packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js

```diff
@@ -1,142 +1,184 @@
+	// 1) limit the amount of requests processed at the same time
+	// 2) save the menu only after all requests are finalized
+	const debouncedBlocks = useDebouncedValue( blocks, 800 );
+	const promiseQueueRef = useRef( new PromiseQueue() );
+	const processedBlocksIds = useRef( [] );
+	useEffect(
+		function() {
+			for ( const clientId of getAllClientIds( debouncedBlocks ) ) {
+				// Already has a related menu item
+				if ( clientId in menuItemsRef.current ) {
+					continue;
 				}
-				const block = createBlockFromMenuItem(
-					item,
-					menuItemInnerBlocks
+				// Already being processed
+				if ( processedBlocksIds.current.includes( clientId ) ) {
+					continue;
+				}
+				processedBlocksIds.current.push( clientId );
+				promiseQueueRef.current.enqueue( () =>
+					createDraftMenuItem( clientId ).then(
+						async ( menuItem ) => {
+							menuItemsRef.current[ clientId ] = menuItem;
+							processedBlocksIds.current.splice(
+								processedBlocksIds.current.indexOf( clientId )
+							);
+						}
+					)
 				);
-				menuItemsRef.current[ block.clientId ] = item;
-				innerBlocks.push( block );
 			}
-			return innerBlocks;
-		};
-
-		// createMenuItemBlocks takes an array of top-level menu items and recursively creates all their innerBlocks
-		const innerBlocks = createMenuItemBlocks( itemsByParentID[ 0 ] );
-		setBlocks( [ createBlock( 'core/navigation', {}, innerBlocks ) ] );
-	}, [ menuItems ] );
-
-	const saveBlocks = () => {
-		const { clientId, innerBlocks } = blocks[ 0 ];
-		const parentItemId = menuItemsRef.current[ clientId ]?.parent;
-
-		const saveNestedBlocks = async ( nestedBlocks, parentId = 0 ) => {
-			for ( const block of nestedBlocks ) {
-				const menuItem = menuItemsRef.current[ block.clientId ];
-				let currentItemId = menuItem?.id || 0;
-
-				if ( ! menuItem ) {
-					const savedItem = await saveMenuItem( {
-						...createMenuItemAttributesFromBlock( block ),
-						menus: menuId,
-						parent: parentId,
-					} );
-					if ( block.innerBlocks.length ) {
-						currentItemId = savedItem.id;
-					}
-				}
+		},
+		[ debouncedBlocks ]
+	);
 
-				if (
-					menuItem &&
-					! isEqual(
-						block.attributes,
-						createBlockFromMenuItem( menuItem ).attributes
-					)
-				) {
-					saveMenuItem( {
-						...menuItem,
-						...createMenuItemAttributesFromBlock( block ),
-						menus: menuId, // Gotta do this because REST API doesn't like receiving an array here. Maybe a bug in the REST API?
-						parent: parentId,
-					} );
-				}
+	// Save handler
+	const { receiveEntityRecords } = useDispatch( 'core' );
+	const { createSuccessNotice, createErrorNotice } = useDispatch(
+		'core/notices'
+	);
 
-				if ( block.innerBlocks.length ) {
-					saveNestedBlocks( block.innerBlocks, currentItemId );
-				}
-			}
-		};
+	const saveBlocks = async () => {
+		const result = await batchSave( menuId, menuItemsRef, blocks[ 0 ] );
+
+		if ( result.success ) {
+			const mapping = menuItemsRef.current;
```

Comment:
what does mapping mean here?

Reply:
We don't need that variable anymore, I just updated the code